File: client/js/party.js
define([], function () {

  var Party = Class.extend({
    init: function (leaderID, membersIDs) {
      this._leader = globalGame.getPlayerByID(leaderID);
      this._members = {};
      this._membersCount = 0;
      this._capacity = 3;
      var players = globalGame.getPlayersByIDs(membersIDs);
      for (var x in players) {
        this.joined(players[x]);
      }
    },

    _add: function (player) {
      this._members[player.id] = player;
      player.party = this;
      this._membersCount++;
    },

    _remove: function (player) {
      delete this._members[player.id];
      player.party = null;
      this._membersCount--;
    },

    joined: function (player) {
      this._add(player);

      if (player != globalGame.player) {
        globalGame.client.notice("%s has joined the party.", player.name);
      }
    },

    left: function (player) {
      this._remove(player);

      if (player == globalGame.player) {
        globalGame.client.notice("You have left the party.");
      } else {
        globalGame.client.notice("%s has left the party.", player.name);
      }
    },

    kicked: function (kicker, kicked) {
      this._remove(kicked);

      if (kicked == globalGame.player) {
        globalGame.client.notice("You were kicked from the party by %s.", kicker.name);
        return;
      }

      globalGame.client.notice(
        "%s kicked %s from the party.",
        kicker.name,
        kicked.name
      );
    },

    setLeader: function (player) {
      this._leader = player;
      if (player == globalGame.player) {
        globalGame.client.notice("You are now the group leader.");
      } else {
        globalGame.client.notice("%s is now the group leader.", player.name);
      }
    },

    getLeader: function () {
      return this._leader;
    },

    getMembers: function () {
      return this._members;
    },

    isFull: function () {
      return this._membersCount == this._capacity;
    },

    isLeader: function (player) {
      return this._leader == player;
    },

    isMember: function (player) {
      return (player.id in this._members);
    }
  });

  return Party;
});

File: client/dart/party.dart
library party;

import "base.dart";
import "game.dart";
import "player.dart";

class Party extends Base {

  Player leader;
  Map<int, Player> members = new Map<int, Player>();
  int size = 0;
  int capacity = 3;

  Party(Player this.leader, List<Player> members) {
    members.forEach((Player member) {
      this.joined(member);
    });
  }

  void _add(Player player) {
    if (this.isMember(player)) {
      throw "${player.id} is already in the party.";
    }

    this.members.putIfAbsent(player.id, () => player);
    player.party = this;
    this.size++;
  }

  void _remove(Player player) {
    if (!this.isMember(player)) {
      throw "${player.id} is not in the party.";
    }

    this.members.remove(player.id);
    player.party = null;
    this.size--;
  }

  void joined(Player player) {
    this._add(player);

    if (player != Game.player) {
      Game.client.notice("${player.name} has joined the party.");
    }
  }

  void left(Player player) {
    this._remove(player);

    if (player == Game.player) {
      Game.client.notice("You have left the party.");
    } else {
      Game.client.notice("${player.name} has left the party.");
    }
  }

  void kicked(Player kicker, Player kicked) {
    this._remove(kicked);

    if (kicked == Game.player) {
      Game.client.notice("You were kicked from the party by ${kicker.name}.");
      return;
    }

    Game.client.notice("${kicker.name} kicked ${kicked.name} from the party.");
  }

  void setLeader(Player player) {
    this.leader = player;
    if (player == Game.player) {
      Game.client.notice("You are now the group leader.");
    } else {
      Game.client.notice("${player.name} is now the group leader.");
    }
  }

  Player getLeader() => this.leader;

  Map<int, Player> getMembers() => this.members;

  bool isFull() => (this.size == this.capacity); 

  bool isLeader(Player player) => (this.leader == player);

  bool isMember(Player player) => this.members.containsKey(player.id);
}
